rclcpp/minimal_action_server/not_composable.cpp

```diff
@@ -0,0 +1,101 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <inttypes.h>
+#include <memory>
+#include "example_interfaces/action/fibonacci.hpp"
+#include "rclcpp/rclcpp.hpp"
+// TODO(jacobperron): Remove this once it is included as part of 'rclcpp.hpp'
+#include "rclcpp_action/rclcpp_action.hpp"
+
+using Fibonacci = example_interfaces::action::Fibonacci;
+using GoalHandleFibonacci = rclcpp_action::ServerGoalHandle<Fibonacci>
+
+
+rclcpp_action::GoalResponse handle_goal(const std::shared_ptr<GoalHandleFibonacci> goal_handle)
+{
+  // Let's reject sequences that are over 9000
+  if (goal_handle->goal.order > 9000)
+  {
+    return rclcpp_action::GoalResponse::REJECT;
+  }
+  return rclcpp_action::GoalResponse::ACCEPT;
+}
+
+rclcpp_action::CancelResponse handle_cancel(const std::shared_ptr<GoalHandleFibonacci> goal_handle)
+{
+  return rclcpp_action::CancelResponse::ACCEPT;
+}
+
+rclcpp_action::ResultResponse<Fibonacci::Result> execute(
+  const std::shared_ptr<GoalHandleFibonacci> goal_handle)
+{
+  rclcpp::Rate loop_rate(1);
+  const auto& goal = goal_handle->goal;
+  auto feedback = Fibonacci::Feedback();
+  auto& sequence = feedback.sequence;
+  sequence.push_back(0);
+  sequence.push_back(1);
+  auto result_response = rclcpp_action::ResultResponse<Fibonacci::Result>();
+
+  for (int i = 1; (i < goal->order) && rclcpp::ok(); ++i)
+  {
+    // Check if there is a cancel request
+    if (goal_handle->is_cancel_request())
+    {
+      result_response.response = rclcpp_action::ResultResponse::CANCELED;
+      result_response.result.sequence = sequence;
+      result_response.message = "Canceling goal at clients request.";
+      return result_response;
+    }
+
+    // Check if goal is done
+    if (i > goal->order)
+    {
+      result_response.response = rclcpp_action::ResultResponse::SUCCEEDED;
+      result_response.result.sequence = sequence;
+      result_response.message = "Successfully computed Fibonacci sequence.";
+      return result_response;
+    }
+    // Update sequence
+    sequence.push_back(sequence[i] + sequence[i - 1]);
+    // Publish feedback
+    goal_handle->publish_feedback(feedback);
+
+    loop_rate.sleep();
+  }
+}
+
+int main(int argc, char ** argv)
+{
+  rclcpp::init(argc, argv);
+  auto node = rclcpp::Node::make_shared("minimal_action_server");
+
+  // Create an action server with three callbacks
+  //   'handle_goal' and 'handle_cancel' are called by the Executor (rclcpp::spin)
+  //   'execute' is called whenever 'handle_goal' returns by accepting a goal
+  //    Calls to 'execute' are made in an available thread from a pool of four.
+  auto action_server = rclcpp_action::create_action_server<Fibonacci>(
+    node,
+    "fibonacci",
+    handle_goal,
+    handle_cancel,
+    execute,
+    rclcpp_action::StaticThreadPoolExecutePolicy(4));
```

Comment:
My understanding is that the signature of the `execute` callback can optionally return `void`, which essentially means you can dispatch any asynchronous tasks and return from it immediately, electing to asynchronously "finish" the execution of the action in the future by using `set_canceled`, `set_aborted`, or `set_success` asynchronously.

In that case the `execute` function is just your notification that the goal is ready to be executed.

Also, if you don't want the asynchronous execute notice at all, one possible value for the "execution policy" is "do nothing". Then it's up to the user to poll the action server object to "drain" ready goals and execute them to one of the terminal states.

So in short, I think what you're asking for should be possible in the current design.

Reply:
👍 Sounds good.

Please forgive a dumb question: Is the contents of the `rclcpp_action/rclcpp_action.hpp` file available somewhere for review so that I might intuit that answer myself?

I also have one more sort-of-pedantic question: The given execution callback does not appear to return a value if the loop aborts early because of `rclcpp::ok()` returning false. In this case, what would the appropriate value be aborted?

Reply:
> Please forgive a dumb question: Is the contents of the rclcpp_action/rclcpp_action.hpp file available somewhere for review so that I might intuit that answer myself?

No unfortunately, we're sort in the process of developing these API's. This pr/examples is meant to be a motivating example for how the API should work. So basically it's a made up API at the moment 😄. I also expect it will evolve more. So for now, asking us questions about the examples is the most reasonable thing.

> I also have one more sort-of-pedantic question: The given execution callback does not appear to return a value if the loop aborts early because of rclcpp::ok() returning false. In this case, what would the appropriate value be aborted?

The user defines the "execute" callback so if in the process they check for and notice the `rclcpp::ok` is returning `false`, then they can choose to call `set_abort` themselves.

If it occurs inbetween calls to the user (e.g. after `handle_goal` has completed but before `execute` is called), I'm not 100% sure right now. I'd assume the action server would cancel all outstanding goals and then wait for any on going user callbacks to finish executing. It's something to maybe keep an eye and ask about again when we have a more complete prototype.